hi-ml-cpath/src/health_cpath/utils/report_utils.py

```diff
@@ -154,12 +166,21 @@ def get_best_epochs(metrics_df: pd.DataFrame, primary_metric: str, maximise: boo
     :param metrics_df: Metrics dataframe, as returned by :py:func:`collect_hyperdrive_metrics()` and
         :py:func:`~health_azure.aggregate_hyperdrive_metrics()`.
     :param primary_metric: Name of the reference metric to optimise.
+    :max_epochs_dict: A dictionary of the maximum number of epochs in each cross-validation round.
     :param maximise: Whether the given metric should be maximised (minimised if `False`).
     :return: Dictionary mapping each hyperdrive child index to its best epoch.
     """
-    best_fn = np.argmax if maximise else np.argmin
-    best_epochs = metrics_df.loc[primary_metric].apply(best_fn)
-    return best_epochs.to_dict()
+    best_epochs = {}
+    for i in metrics_df.columns:
+        primary_metric_list = metrics_df[i][primary_metric]
+        if primary_metric_list is not None:
+            # If extra validation epoch was logged (N+1), return only the first N elements
+            primary_metric_list = primary_metric_list[:-1] \
+                if (len(primary_metric_list) == max_epochs_dict[i] + 1) else primary_metric_list
```

Comment:
Only if we have more metric elements than the maximum number of epochs (e.g., when we have an extra validation epoch), it will drop the last element. Otherwise, the full list should be considered. Could you please explain the comment a bit more?